internal/status/status.go

```diff
@@ -154,12 +154,12 @@ func (s *Status) Details() []any {
 	}
 	details := make([]any, 0, len(s.s.Details))
 	for _, any := range s.s.Details {
-		detail := &ptypes.DynamicAny{}
-		if err := ptypes.UnmarshalAny(any, detail); err != nil {
+		detail, err := any.UnmarshalNew()
+		if err != nil {
 			details = append(details, err)
 			continue
 		}
-		details = append(details, detail.Message)
+		details = append(details, detail)
```

Comment:
Are these new-style protos compatible with users running with `github.com/golang/proto` and `ptypes`?  If not, we can't make this part of this change.

Reply:
How would I go about testing this? Should I write a `status_test.go` and check that both versions work?

Reply:
For example, I wrote this:

```go
func (s) TestWithDetailsVersionsCompat(t *testing.T) {
	details := []protoreflect.ProtoMessage{
		ptypes.TimestampNow(),
		ptypes.DurationProto(1 * time.Hour),
		&timestamppb.Timestamp{},
		&durationpb.Duration{},
	}
	status := New(codes.InvalidArgument, "an error")

	if _, err := status.WithDetails(details...); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}
```

and it passes.

Reply:
Not sure about how to test DynamicAny though, if you have any idea let me know.

Reply:
@dfawley

Reply:
> How would I go about testing this? Should I write a status_test.go and check that both versions work?

Yes, that sounds good.

> For example, I wrote this [...] and it passes.

This doesn't look like what I was thinking (because it's testing `WithDetails`, not `Details`), but it's still a useful test.  What I had in mind was something that called `Details` on a status containing details, and does type assertions using protov1 messages:

```go
st := status.New().WithDetails(ptypes.TimestampNow(), timestamppb.Now()) // Create with old ptypes and new tspb
x := st.Details()[0]
if _, ok := x.(protov1.Message); !ok {
	t.Fatal()
}
if _, ok := x.(timestamppb.Timestamp); !ok {
	t.Fatalf()
}
```

I thought that `anypb.UnmarshalNew()` produced _new_ `protov2.Message`s which were incompatible with _old_ `protov1.Message`s.  But if the new `protov2.Message` is a superset of the old `protov1.Message` then I think this should be safe to change, especially because it's returned as `any` instead of `protov1.Message`.